### PR TITLE
Feature/dan uncommitted logic

### PR DESCRIPTION
**Pull request by qualidan**
added two functions to utils.py
switch_to_temp_branch
revert_from_temp_branch

Both are called in case repo is dirty (both from bp and sb)

added basic initial unit tests

### CODE REVIEW

File: colony/commands/bp.py [hunk 1]
```diff
@@ -38,34 +46,57 @@ def do_validate(self):
...
                 self.message(f"Automatically detected current working branch: {working_branch}")
+
             except BadBlueprintRepo as e:
                 working_branch = None
                 logger.warning(
                     f"No branch has been specified and it could not be identified from the working directory; "
                     f"reason: {e}. A branch of the Blueprints Repository attached to Colony Space will be used"
                 )
 
+            if not remote and not repo.is_current_branch_synced():
```
Comment (ddovbii) [inline colony/commands/bp.py:74]: repo and working_branch here could None or undefined

**Review by ddovbii — CHANGES_REQUESTED** [review]

File: colony/commands/bp.py (hunk 1 above)
Comment (qualidan) [reply colony/commands/bp.py:74]: I think this should solve it:
if not remote and repo and not repo.is_current_branch_synced():
but might have other implications

Comment (qualidan) [reply colony/commands/bp.py:74]: double checking I think I meant:
if not remote and working_branch and not repo.is_current_branch_synced():

File: colony/commands/bp.py [hunk 2]
```diff
@@ -38,34 +46,61 @@ def do_validate(self):
...
             except BadBlueprintRepo as e:
                 working_branch = None
                 logger.warning(
                     f"No branch has been specified and it could not be identified from the working directory; "
                     f"reason: {e}. A branch of the Blueprints Repository attached to Colony Space will be used"
                 )
 
+            # Checking if:
+            # 1) User has specified not use local (not remote)
+            # 2) User is in ana actual git dir (working_branch)
```
Comment (alexazarh) [inline colony/commands/bp.py:76]: there is a typo in the comment

File: colony/commands/sb.py [hunk 3]
```diff
@@ -48,6 +50,8 @@ class SandboxesCommand(BaseCommand):
 
        -w, --wait <timeout>             Set the timeout in minutes for the sandbox to become active. If not set, command
                                         will not block terminal and just return the ID of started sandbox
+
+       -r, --remote                     Work with remote branch (default is local branch)
```
[the remark this answers is not in the document]
Comment (alexazarh) [reply colony/commands/sb.py:58]: Agree

File: colony/commands/sb.py [hunk 4]
```diff
@@ -111,20 +120,39 @@ def do_start(self):
...
+            if not remote and working_branch and not repo.is_current_branch_synced():
+                temp_working_branch = working_branch
+                try:
+                    temp_working_branch = switch_to_temp_branch(repo, working_branch)
+                except Exception as e:
+                    logger.error("Was not able to create temp branch.")
+                    if working_branch:
+                        logger.warning(f"Branch {working_branch} will be used. Reason: {str(e)}")
+                    else:
+                        logger.warning(f"Remote branch will be used. Reason: {str(e)}")
```
Comment (alexazarh) [inline colony/commands/sb.py:153]: this looks to be a duplicate with what we have in bp.py.
It will good to call the same code from both commands

**Review by alexazarh — CHANGES_REQUESTED** [review]

File: colony/commands/sb.py [hunk 5]
```diff
@@ -111,20 +117,9 @@ def do_start(self):
...
-                self.message(f"Automatically detected current working branch: {working_branch}")
-            except BadBlueprintRepo as e:
-                working_branch = None
-                logger.warning(
-                    f"No branch has been specified and it could not be identified from the working directory; "
-                    f"reason: {e}. A branch of the Blueprints Repository attached to Colony Space will be used"
-                )
-
-        # TODO(ddovbii): This obtaining default values magic mast be refactored
+        repo, working_branch, temp_working_branch = figure_out_branch(branch, name)
```
Comment (ddovbii) [inline colony/commands/sb.py:120]: name -> bp_name

File: colony/utils.py [hunk 6]
```diff
@@ -150,3 +182,100 @@ def parse_comma_separated_string(params_string: str = None) -> dict:
...
+                f"reason: {e}. A branch of the Blueprints Repository attached to Colony Space will be used"
+            )
+
+        # Checking if:
+        # 1) User has specified not use local (specified a branch)
+        # 2) User is in an actual git dir (working_branch)
+        # 3) There is even a need to create a temp branch for out-of-sync reasons:
+        #   either repo.is_dirty() (changes have not been committed locally)
+        #   or repo.is_current_branch_synced() (changes committed locally but not pushed to remote)
+        if not branch and working_branch and (repo.is_dirty() or repo.is_current_branch_synced()):
```
Comment (ddovbii) [inline colony/utils.py:214]: I think it should be `not repo.is_current_branch_synced()`

File: colony/utils.py [hunk 7]
```diff
@@ -150,3 +182,100 @@ def parse_comma_separated_string(params_string: str = None) -> dict:
...
+            try:
+                temp_working_branch = switch_to_temp_branch(repo, working_branch)
+                BaseCommand.message(f"Validating using temp branch: {temp_working_branch}")
+            except Exception as e:
+                logger.warning("Was not able push your latest changes to temp branch for validation.")
+                if working_branch:
```
Comment (ddovbii) [inline colony/utils.py:220]: why do we need this condition here? According to previous `if` condition `working_branch` is always instantiated at this point

File: colony/utils.py [hunk 8]
```diff
@@ -150,3 +182,96 @@ def parse_comma_separated_string(params_string: str = None) -> dict:
...
+                f"reason: {e}. A branch of the Blueprints Repository attached to Colony Space will be used"
+            )
+
+        # Checking if:
+        # 1) User has specified not use local (specified a branch)
+        # 2) User is in an actual git dir (working_branch)
+        # 3) There is even a need to create a temp branch for out-of-sync reasons:
+        #   either repo.is_dirty() (changes have not been committed locally)
+        #   or not repo.is_current_branch_synced() (changes committed locally but not pushed to remote)
+        if not branch and working_branch and (repo.is_dirty() or not repo.is_current_branch_synced()):
```
Comment (ddovbii) [inline colony/utils.py:214]: Unfortunately this condition still has a couple of things we need to address:

1. If a user is working in a local branch which doesn't exist in remote , this block will be unreachable (`working_branch is None`). However, it's a valid case for _temporary_branch_ feature

2. It turned out that is_dirty() checks only changes in already existing and tracked files. But if user has added some new files but hasn't tracked them (`git add file`) -> `is_dirty()` will be `false`. I think it's a valid case for temp branch creation as well.
So, in my opinion the reasons why we want to have a temp branch are the following:
  - you are working in the branch which exists on remote but branches are not synced. **covered**
  - you are working in a branch which doesn't exist on remote
  - repo is dirty. There are new tracked files or changes in existing files. **covered**
  - repo has untracked files. there is a property to check it `repo.untracked_files`

File: colony/utils.py [hunk 9]
```diff
@@ -150,3 +187,93 @@ def parse_comma_separated_string(params_string: str = None) -> dict:
...
+                f"reason: {e}. A branch of the Blueprints Repository attached to Colony Space will be used"
+            )
+
+        # Checking if:
+        # 1) User has specified not use local (specified a branch)
+        # 2) User is in an actual git dir (working_branch)
+        # 3) There is even a need to create a temp branch for out-of-sync reasons:
+        #   either repo.is_dirty() (changes have not been committed locally)
+        #   or not repo.is_current_branch_synced() (changes committed locally but not pushed to remote)
+        if not branch and working_branch and not repo.is_current_state_synced_with_remote:
```
Comment (ddovbii) [inline colony/utils.py:219]: missed brackets

**Review by ddovbii — CHANGES_REQUESTED** [review]

**Pull request merged**